File: src/glcore/mincostflow.cpp
// Integral min-cost flow by successive shortest paths with Johnson potentials [Prop 5.4] (paper_notes §5).
//
// Network arcs are stored in pairs: arc id `a` (even) is the forward arc returned by add_arc, `a ^ 1` is its
// residual partner (capacity 0, cost -cost). `cap` of an E is the *residual* capacity, so the flow on a
// forward arc equals the residual capacity of its partner. Costs must be nonnegative, so the initial
// potentials are zero and every Dijkstra runs on nonnegative reduced costs c(u,v) + h(u) - h(v). Each
// augmentation pushes the bottleneck of the path (capped by the remaining requirement), so all flows stay
// integral ("Must be integral (SSP on integral data is)", paper_notes §5).
#include "mincostflow.hpp"

#include <limits>
#include <queue>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

namespace glcore {

// [Prop 5.4] empty network on num_nodes nodes.
MinCostFlow::MinCostFlow(int num_nodes) : n_(num_nodes), adj_(num_nodes > 0 ? num_nodes : 0), potential_(num_nodes > 0 ? num_nodes : 0, 0) {
    if (num_nodes <= 0) throw std::invalid_argument("MinCostFlow: num_nodes must be positive, got " + std::to_string(num_nodes));
}

// [Prop 5.4] add the arc u -> v with capacity cap >= 0 and cost >= 0; returns its (even) arc id.
int MinCostFlow::add_arc(int u, int v, int64_t cap, int64_t cost) {
    if (u < 0 || u >= n_ || v < 0 || v >= n_)
        throw std::invalid_argument("MinCostFlow::add_arc: node out of range (" + std::to_string(u) + "," + std::to_string(v) +
                                    ") for " + std::to_string(n_) + " nodes");
    if (cap < 0) throw std::invalid_argument("MinCostFlow::add_arc: capacities must be nonnegative");
    if (cost < 0) throw std::invalid_argument("MinCostFlow::add_arc: costs must be nonnegative (zero initial potentials)");
    if (edges_.size() > size_t(std::numeric_limits<int>::max() - 2))
        throw std::length_error("MinCostFlow::add_arc: too many arcs");
    const int id = (int)edges_.size();
    edges_.push_back(E{v, cap, cost});
    adj_[u].push_back(id);
    edges_.push_back(E{u, 0, -cost});
    adj_[v].push_back(id + 1);
    return id;
}

// Flow on arc `arc`: for a forward arc the residual capacity of its partner; for a partner the negative of
// the forward flow (same convention as the reference glref.mincostflow.MinCostFlow.flow).
int64_t MinCostFlow::flow(int arc) const {
    if (arc < 0 || arc >= (int)edges_.size()) throw std::out_of_range("MinCostFlow::flow: arc id out of range");
    return (arc & 1) ? -edges_[arc].cap : edges_[arc ^ 1].cap;
}

// [Prop 5.4] successive shortest paths: repeat { Dijkstra on reduced costs; update potentials; augment by the
// bottleneck } until `required` units are sent or z is unreachable. Returns (value, cost); value < required
// means no flow of the requested value exists (FESAC fails in the split-assignment use).
std::pair<int64_t, int64_t> MinCostFlow::solve(int s, int z, int64_t required) {
    if (s < 0 || s >= n_ || z < 0 || z >= n_) throw std::invalid_argument("MinCostFlow::solve: source/sink out of range");
    if (s == z) throw std::invalid_argument("MinCostFlow::solve: source and sink must differ");
    if (required < 0) throw std::invalid_argument("MinCostFlow::solve: required must be nonnegative");
    const int64_t INF = std::numeric_limits<int64_t>::max();
    std::vector<int64_t> dist(n_);
    std::vector<int> parent_arc(n_);
    // Binary heap keyed by (tentative distance, node): deterministic pop order, lazy deletion.
    using Item = std::pair<int64_t, int>;
    std::priority_queue<Item, std::vector<Item>, std::greater<Item>> pq;
    int64_t value = 0, total_cost = 0;
    while (value < required) {
        // ---- Dijkstra with reduced costs (Johnson potentials keep them nonnegative).
        std::fill(dist.begin(), dist.end(), INF);
        std::fill(parent_arc.begin(), parent_arc.end(), -1);
        dist[s] = 0;
        pq.push(Item{0, s});
        while (!pq.empty()) {
            const Item it = pq.top();
            pq.pop();
            const int u = it.second;
            if (it.first > dist[u]) continue;  // stale entry
            const int64_t du = dist[u];
            for (int a : adj_[u]) {
                const E& e = edges_[a];
                if (e.cap <= 0) continue;
                const int64_t rc = e.cost + potential_[u] - potential_[e.to];  // >= 0 by the potential invariant
                if (rc < 0) throw std::logic_error("MinCostFlow::solve: negative reduced cost; potential invariant violated");
                const int64_t nd = du + rc;
                if (nd < dist[e.to]) {
                    dist[e.to] = nd;
                    parent_arc[e.to] = a;
                    pq.push(Item{nd, e.to});
                }
            }
        }
        if (dist[z] == INF) break;  // maximum flow reached below `required`
        // ---- potentials: h(v) += dist(v) for reachable v keeps every residual reduced cost nonnegative.
        for (int v = 0; v < n_; ++v)
            if (dist[v] != INF) potential_[v] += dist[v];
        // ---- bottleneck along the parent arcs (capped by the remaining requirement).
        int64_t amt = required - value;
        int64_t path_cost = 0;
        for (int v = z; v != s;) {
            const int a = parent_arc[v];
            const E& e = edges_[a];
            if (e.cap < amt) amt = e.cap;
            path_cost += e.cost;
            v = edges_[a ^ 1].to;
        }
        if (amt <= 0) throw std::logic_error("MinCostFlow::solve: zero bottleneck on a shortest path");
        for (int v = z; v != s;) {
            const int a = parent_arc[v];
            edges_[a].cap -= amt;
            edges_[a ^ 1].cap += amt;
            v = edges_[a ^ 1].to;
        }
        value += amt;
        int64_t inc;
        if (__builtin_mul_overflow(amt, path_cost, &inc) || __builtin_add_overflow(total_cost, inc, &total_cost))
            throw std::overflow_error("MinCostFlow::solve: total cost overflows int64");
    }
    return {value, total_cost};
}

}  // namespace glcore

File: reference/glref/mincostflow.py
"""Integral min-cost flow by successive shortest paths (reference implementation).

Used by :func:`glref.weighted.min_cost_split_assignment` [Prop 5.4] of
docs/paper_notes.md §5. All arc costs of that network are small nonnegative
integers (``ξ_v(t) ≤ k``), so the transparent choice is *successive shortest
paths* with a Bellman–Ford/SPFA shortest-path search on the residual graph for
every augmentation (residual arcs carry negative costs, which SPFA handles;
Dijkstra with potentials would be the optimised alternative). Augmentation is by
the bottleneck of the path, so all flows stay integral (paper_notes §5:
"Must be integral (SSP on integral data is)").
"""
from __future__ import annotations

from collections import deque

INF = float("inf")


class MinCostFlow:
    """Adjacency-list min-cost flow network with paired residual arcs.

    ``add_arc(u, v, cap, cost)`` returns the index ``idx`` of the forward arc;
    its residual partner is ``idx ^ 1`` (capacity 0, cost ``-cost``).
    """

    def __init__(self, num_nodes: int) -> None:
        self.num_nodes = num_nodes
        self.head: list[int] = []
        self.cap: list[int] = []
        self.cost: list[int] = []
        self._flow: list[int] = []
        self.adj: list[list[int]] = [[] for _ in range(num_nodes)]

    # ------------------------------------------------------------ building
    def add_arc(self, u: int, v: int, cap: int, cost: int) -> int:
        if cap < 0:
            raise ValueError("capacities must be nonnegative")
        idx = len(self.head)
        self.head.append(v)
        self.cap.append(cap)
        self.cost.append(cost)
        self._flow.append(0)
        self.adj[u].append(idx)
        self.head.append(u)
        self.cap.append(0)
        self.cost.append(-cost)
        self._flow.append(0)
        self.adj[v].append(idx + 1)
        return idx

    def flow(self, idx: int) -> int:
        """Flow currently on arc ``idx`` (negative on residual partners)."""
        return self._flow[idx]

    def residual(self, a: int) -> int:
        return self.cap[a] - self._flow[a]

    # ------------------------------------------------------------ solving
    def _shortest_path(self, s: int, z: int) -> list[int] | None:
        """SPFA (queue-based Bellman–Ford) on the residual graph.

        Returns ``parent_arc`` (arc index entering each node) or ``None`` when
        ``z`` is unreachable. The residual graph never contains a negative cycle
        along a successive-shortest-path run started from the zero flow with
        nonnegative costs, so SPFA terminates.
        """
        n = self.num_nodes
        dist: list[float] = [INF] * n
        parent: list[int] = [-1] * n
        in_queue = [False] * n
        dist[s] = 0
        dq: deque[int] = deque([s])
        in_queue[s] = True
        relaxations = 0
        limit = n * len(self.head) + 1  # negative-cycle guard (cannot trigger in SSP)
        while dq:
            u = dq.popleft()
            in_queue[u] = False
            du = dist[u]
            for a in self.adj[u]:
                if self.cap[a] - self._flow[a] <= 0:
                    continue
                v = self.head[a]
                nd = du + self.cost[a]
                if nd < dist[v]:
                    dist[v] = nd
                    parent[v] = a
                    relaxations += 1
                    if relaxations > limit:
                        raise RuntimeError("negative cycle detected in residual graph")
                    if not in_queue[v]:
                        in_queue[v] = True
                        dq.append(v)
        if dist[z] == INF:
            return None
        return parent

    def min_cost_flow(self, s: int, z: int, required: int) -> tuple[int, int]:
        """Push up to ``required`` units from ``s`` to ``z`` at minimum cost.

        Successive shortest paths, each augmented by its bottleneck (capped at
        the remaining requirement). Returns ``(value, cost)``; ``value <
        required`` means no flow of the requested value exists.
        """
        if s == z:
            raise ValueError("source and sink must differ")
        value = 0
        total_cost = 0
        while value < required:
            parent = self._shortest_path(s, z)
            if parent is None:
                break
            # bottleneck along the path
            amt = required - value
            v = z
            while v != s:
                a = parent[v]
                r = self.cap[a] - self._flow[a]
                if r < amt:
                    amt = r
                v = self.head[a ^ 1]
            assert amt > 0
            v = z
            path_cost = 0
            while v != s:
                a = parent[v]
                self._flow[a] += amt
                self._flow[a ^ 1] -= amt
                path_cost += self.cost[a]
                v = self.head[a ^ 1]
            value += amt
            total_cost += amt * path_cost
        return value, total_cost

    def total_cost(self) -> int:
        """Cost of the current flow, ``Σ flow(a) · cost(a)`` over forward arcs."""
        return sum(self._flow[a] * self.cost[a] for a in range(0, len(self.head), 2))
